apps/src/javalab/Javalab.js

```diff
@@ -120,7 +139,117 @@ Javalab.prototype.onContinue = function() {
+    return;
+  }
+  header.showProjectSaving();
+  filesApi.renameFile(
+    filename,
+    newFilename,
+    () => {
+      // at this point project.filesVersionId will be updated and we can save
+      // sources as well.
+      project.save();
+    },
+    () => {
+      console.warn(`Javalab: error file ${filename} not renamed`);
+      project.showSaveError_();
+    }
+  );
 };
 
+Javalab.prototype.onSave = function() {
+  header.showProjectSaving();
+  const filename = getStore().getState().javalab.filename;
+  const editorText = getStore().getState().javalab.editorText;
+  // TODO: enable multi-file
+  filesApi.putFile(
+    filename,
+    editorText,
+    () => {
+      // at this point project.filesVersionId will be updated and we can save
+      // sources as well.
+      project.save();
+    },
+    () => {
+      console.warn(`Javalab: error file ${filename} not renamed`);
+      project.showSaveError_();
+    }
+  );
+};
+
+/**
+ * Load the file entry list and store it as this.fileEntries
+ */
+Javalab.prototype.loadFiles = function(callback) {
+  const onFilesReady = (files, filesVersionId) => {
+    assetListStore.reset(files);
+    const fileEntries = assetListStore.list().map(fileEntry => ({
+      name: fileEntry.filename,
+      url: filesApi.basePath(fileEntry.filename),
+      versionId: fileEntry.versionId
+    }));
+    this.populateFiles(fileEntries, callback);
+    this.initialFilesVersionId = this.initialFilesVersionId || filesVersionId;
+
+    if (filesVersionId !== this.initialFilesVersionId) {
+      // After we've detected the first change to the version, we store this
+      // version id so that subsequent writes will continue to replace the
+      // current version (until the browser page reloads)
+      project.filesVersionId = filesVersionId;
+    }
+  };
+
+  filesApi.getFiles(
+    result => onFilesReady(result.files, result.filesVersionId),
+    xhr => {
+      if (xhr.status === 404) {
+        // No files in this project yet, proceed with an empty file
+        // list and no start version id
+        onFilesReady([], null);
+      } else {
+        console.error('files API failed, status: ' + xhr.status);
+      }
+    },
+    this.getCurrentFilesVersionId()
+  );
+};
+
+Javalab.prototype.populateFiles = function(fileEntries, callback) {
+  if (fileEntries.length === 0) {
+    // TODO: make starter code more customizable
+    callback();
+  } else {
+    // TODO: enable multi-file
+    const fileEntry = fileEntries[0];
+    requestFileEntryAndWrite(fileEntry, callback);
+  }
+};
+
+function requestFileEntryAndWrite(fileEntry, callback) {
+  // read the data
+  $.ajax(`${fileEntry.url}?version=${fileEntry.versionId}`, {
+    dataType: 'text'
+  })
+    .done(fileContent => {
+      getStore().dispatch(setFileName(fileEntry.name));
+      getStore().dispatch(setEditorText(fileContent));
+      callback();
+    })
+    .fail((jqXHR, textStatus, errorThrown) => {
+      callback();
```

Comment:
right now, `<JavalabView/>` will still render as normal if we fail here. this may be future work, but what should happen if this fails? one idea would be to pass an error through to `<JavalabView/>` and render an error page instead of the normal view to let the user know their file couldn't be loaded.

Reply:
good call, I think I missed that.